lisa/tests/scheduler/load_tracking.py

```diff
@@ -587,35 +570,49 @@ def test_freq_invariance(self) -> ResultBundle:
 
         .. seealso:: :class:`InvarianceItem.test_task_util_avg`
         """
+
         logger = self.get_logger()
-        metrics = {}
-        passed = False
-        for cpu, item_group in groupby(self.invariance_items, key=lambda x: x.cpu):
-            group_passed = True
-            freq_list = []
-            for item in item_group:
-                freq_list.append(item.freq)
-                # Only test util, as it should be more robust
-                res = item.test_task_util_avg()
-                passed &= bool(res)
-                name = '{}@{}'.format(cpu, item.freq)
-                metrics[name] = res.metrics
 
-            logger.info('Util avg invariance {res} for CPU {cpu} at frequencies: {freq_list}'.format(
-                res='passed' if group_passed else 'failed',
+        def make_group_bundle(cpu, item_group):
+            bundle = AggregatedResultBundle(
+                [
+                    # Only test util, as it should be more robust
+                    item.test_task_util_avg()
+                    for item in item_group
+                ],
+                # each item's "cpu" metric also contains the frequency
+                name_metric='cpu',
+            )
+            # At that level, we only report the CPU, since nested bundles cover
+            # different frequencies
+            bundle.add_metric('cpu', cpu)
+
+            logger.info('Util avg invariance {res} for CPU {cpu}'.format(
+                res=bundle.result.lower_name,
                 cpu=cpu,
-                freq_list=freq_list,
             ))
+            return bundle
 
-            # At least one group must pass
-            passed |= group_passed
-
-        res = ResultBundle.from_bool(passed)
-        for cpu, submetric in metrics.items():
-            res.add_metric(cpu, submetric)
-
-        return res
+        group_result_bundles = [
+            make_group_bundle(cpu, item_group)
+            for cpu, item_group in groupby(self.invariance_items, key=lambda x: x.cpu)
+        ]
 
+        # The combination differs from the AggregatedResultBundle default one:
```

Comment:
I know this isn't part of this change, but I didn't realise/forgot this test behaved this way. Why don't we use the default combination method?

Reply:
The logic of the test itself makes it pass as long as at least one CPU is frequency invariant. If the invariance is broken on one CPU but not the other one, it's probably a cpu invariance issue. I say "probably" because at the end of the day, there is no way to tell apart a freq invariance issue from a cpu invariance issue with what we can observe. Both tests are just heuristics to guide the debugging, but ultimately cannot be relied upon. The only one true source of information are the (cpu, freq) tests, which should be looked at one by one if something goes wrong to figure out what is broken.